core/services/relay/evm/read/multieventtype.go

```diff
@@ -0,0 +1,156 @@
+package read
+
+import (
+	"context"
+	"fmt"
+	"iter"
+	"reflect"
+	"strconv"
+	"strings"
+
+	"github.com/ethereum/go-ethereum/common"
+
+	commontypes "github.com/smartcontractkit/chainlink-common/pkg/types"
+	"github.com/smartcontractkit/chainlink-common/pkg/types/query"
+	"github.com/smartcontractkit/chainlink-common/pkg/values"
+	"github.com/smartcontractkit/chainlink/v2/core/chains/evm/logpoller"
+)
+
+type EventQuery struct {
+	Filter           query.KeyFilter
+	EventBinding     *EventBinding
+	SequenceDataType any
+	IsValuePtr       bool
+	Address          common.Address
+}
+
+func MultiEventTypeQuery(ctx context.Context, lp logpoller.LogPoller, eventQueries []EventQuery, limitAndSort query.LimitAndSort) (sequences iter.Seq2[string, commontypes.Sequence], err error) {
+	defer func() {
+		if err != nil {
+			if len(eventQueries) > 0 {
+				var calls []Call
+				for _, eq := range eventQueries {
+					calls = append(calls, Call{
+						ContractAddress: eq.Address,
+						ContractName:    eq.EventBinding.contractName,
+						ReadName:        eq.EventBinding.eventName,
+						ReturnVal:       eq.SequenceDataType,
+					})
+				}
+
+				err = newErrorFromCalls(err, calls, "", eventReadType)
+			} else {
+				err = fmt.Errorf("no event queries provided: %w", err)
+			}
+		}
+	}()
+
+	for _, eq := range eventQueries {
+		if err = eq.EventBinding.validateBound(eq.Address); err != nil {
+			return nil, err
+		}
+	}
+
+	allFilterExpressions := make([]query.Expression, 0, len(eventQueries))
+	for _, eq := range eventQueries {
+		var expressions []query.Expression
+
+		defaultExpressions := []query.Expression{
+			logpoller.NewAddressFilter(eq.Address),
+			logpoller.NewEventSigFilter(eq.EventBinding.hash),
+		}
+		expressions = append(expressions, defaultExpressions...)
+
+		remapped, remapErr := eq.EventBinding.remap(eq.Filter)
+		if remapErr != nil {
+			return nil, fmt.Errorf("error remapping filter: %w", err)
+		}
+		expressions = append(expressions, remapped.Expressions...)
+
+		filterExpression := query.And(expressions...)
+
+		allFilterExpressions = append(allFilterExpressions, filterExpression)
+	}
+
+	eventQuery := query.Or(allFilterExpressions...)
+
+	queryName := ""
```

Comment:
Could this string get long enough to cause problems for prometheus metrics label since that is what this is used ro?

Reply:
It doesn't need to included the contract name and address for each event name, it could unique that, I'll make that change

Reply:
changed this as described to shorten the name